src/block/hashtable.rs

```diff
@@ -197,14 +197,16 @@ impl HashTable for HashTableU16 {
 #[inline]
 pub fn get_table_size(input_len: usize) -> (usize, usize) {
     let (dict_size, dict_bitshift) = match input_len {
+        // U16 Positions
         0..=500 => (128, 9),
         501..=1_000 => (256, 8),
         1_001..=4_000 => (512, 7),
         4_001..=8_000 => (1024, 6),
         8_001..=16_000 => (2048, 5),
-        16_001..=30_000 => (8192, 3),
-        // 100_000..=400_000 => (8192, 3),
-        _ => (16384, 2),
+        16_001..=65535 => (8192, 3),
+        // U32 positions => 16KB table
+        // Usize (U64) positions => 32KB table
+        _ => (4096, 4),
```

Comment:
the hash table shrinks for sizes >65K

Reply:
It's intentional (as per commit msg), these were much larger then the reference implementation which is  16k bytes total.

Reply:
I mean the hashtable is larger for input sizes 16_001..=65535 than input sizes >65535.

In the block case for very small data (1KB) it was faster with a small table size if I remember correctly. I'm not sure how allocating a hashtable with 4k entries is faster than one with 256.

This also affects compression ratio, did you measure it?

Reply:
Ah I just saw the size wasn't forwarded anyway

Reply:
> This also affects compression ratio, did you measure it?

It does, yes. I don't have the numbers anymore though.

> Ah I just saw the size wasn't forwarded anyway

Exactly. So the U32 table was being 64KB total.

It should be possible to make the table size customizeable, but since the reference implementation never did (except in compile time) I doubt it's worth the hassle :shrug:

Reply:
Still, having a smaller hashtable for larger data doesn't make sense imo

```
16_001..=65535 => (8192, 3),
_ => (4096, 4),
```

Reply:
I see your point. Do you know how the previous numbers chosen? And what do you think it should ultimately look like?

I imagine the smaller tables for smaller inputs resulted in a speed up since it was faster to fill with zeroes, at little compression ratio cost. But if we're not sure I'd ratter settle for a behavior similar to the reference implementation which should have more empirical data behind it.

Either way this is only for the bloc compressor, which always knows the size of the input. The FrameCompressor, where matters the most, can continue to use the fixed 16KB table (as per reference).

Reply:
I suggest something like this, which is still arbitrary but still grows towards the 16KB limit faster.

```rust
pub fn get_table_size(input_len: usize) -> (usize, usize) {
    let (dict_size, dict_bitshift) = match input_len {
        // U16 Positions
        0..=65535 => {
            // A table up to 16K bytes in size, considering each slot takes 2 bytes.
            let size = input_len.next_power_of_two().min(16 * 1024) / 2;
            (size, 16 - size.trailing_zeros() as usize)
        }
        // U32 positions => 16KB table
        // Usize (U64) positions => 32KB table
        _ => (4096, 4),
    };
    (dict_size, dict_bitshift)
}

```

Reply:
I did some measurements on speed and compression and interpolated some values.
The scenario would be a service which receives compressed data of mostly small data. It would be wasteful to allocate 16kb for e.g. 1kb packets.

Such a function would be fine for me, but we should test for all sizes, that we get correct pairs. When the bitshift is wrong, we would waste size in the hashmap or it would fail.

Reply:
Adjusted in d197a09